rustls/src/server/verifier_builder.rs

```diff
@@ -0,0 +1,318 @@
+use std::sync::Arc;
+
+use crate::server::UnparsedCertRevocationList;
+use crate::verify::{
+    AnonymousClientsPolicy, ClientCertVerifier, NoClientAuth, WebpkiClientVerifier,
+};
+use crate::{CertRevocationListError, RootCertStore};
+
+/// Possible error states produced by [`ClientCertVerifierBuilder::build`].
+#[derive(Debug, Clone, Eq, PartialEq)]
+#[non_exhaustive]
+pub enum ClientCertVerifierBuilderError {
+    /// The builder was configured with CRLs, but no trust anchors. Client authentication
+    /// would not be performed, making the CRLs useless.
+    CrlsWithoutRoots,
+    /// The builder was configured to reject unauthenticated clients but no trust anchors were
+    /// provided. Client authentication would not be performed, allowing anonymous clients
+    /// in contradiction to the builder configuration.
+    RequiredAuthWithoutRoots,
+}
+
+/// A builder for configuring a `webpki` client certificate verifier.
+///
+/// For more information, see the [`WebpkiClientVerifier`] documentation.
+#[derive(Debug, Clone)]
+pub struct ClientCertVerifierBuilder {
+    roots: Option<RootCertStore>,
+    crls: Vec<webpki::OwnedCertRevocationList>,
+    anon_policy: AnonymousClientsPolicy,
+}
+
+impl ClientCertVerifierBuilder {
+    pub(crate) fn new() -> Self {
+        Self {
+            roots: None,
+            crls: Vec::new(),
+            anon_policy: AnonymousClientsPolicy::Forbidden,
+        }
+    }
+
+    /// Verify that if a client offers a client certificate, that it is issued by a trust anchor
+    /// present in the provided [RootCertStore]. Calling `with_roots` multiple times will merge
+    /// the trust anchors.
+    ///
+    /// All clients must provide a client certificate unless you have allowed unauthenticated
+    /// clients by calling [allow_unauthenticated()][ClientCertVerifierBuilder::allow_unauthenticated]
+    /// on the builder.
+    pub fn with_roots(self, roots: RootCertStore) -> Self {
+        Self {
+            roots: match self.roots {
+                None => Some(roots),
+                Some(mut existing) => {
+                    existing.add_server_trust_anchors(roots.roots.into_iter());
```

Comment:
Given `RootCertStore` is actually agnostic of server/client, I think this function is now misnamed?

Reply:
(er, this isn't really a result of this PR. just a general observation)

Reply:
That's a good point. I tacked on a second commit to rename this function: 41d25d4fb1e74104f6e0efa2cc4fd15ab499674e